documentation/client-side-forms/tutorial-binder-validation.asciidoc

```diff
@@ -0,0 +1,358 @@
+The `Binder` API validates most standard constraints such as `@Max`, `@Size`, `@Pattern`, `@Email` on the client-side, without a network round trip delay
+(see the full list in the <<tutorial-binder-validation#built-in-client-side-validators>> section below).
+When you eventually submit data to server-side endpoints, Vaadin validates all constraints on the server as well, and the `Binder` API would update the form to show server-side validation errors (if any).
+
+This chapter describes input validation for forms created with TypeScript and LitElement (the `@vaadin/form` module).
+For details on input validation for forms created in Java see the <<../binding-data/tutorial-flow-components-binder-validation#,Validating and Converting User Input>> chapter.
+
+=== Specifying Constraints
+
+Constraints are specified as a part of the data model, in the Java code.
+You can use any of the https://beanvalidation.org/2.0/spec/#builtinconstraints[built-in constraints] as well as your own.
+
+.Defining constraints on individual object properties using the Bean Validation annotations
+[source, java]
+----
+import javax.validation.constraints.Email;
+import javax.validation.constraints.NotBlank;
+
+public class Employee {
+    @NotBlank
+    private String username;
+
+    private String title;
+
+    @Email(message = "Please enter a valid e-mail address")
+    private String email;
+
+    // + other fields, constructors, setters and getters
+}
+----
+
+During the build, when Vaadin generates TypeScript types, it includes the constraint information into the generated model types.
+All forms working with the same entity type have the same set of constraints.
+
+When you create a form for an entity type it gets the user input validation automatically.
+
+.Binding form fields to the data model
+[source, typescript]
+----
+import { Binder, field } from '@vaadin/form';
+import EmployeeModel from './generated/com/example/application/EmployeeModel';
+
+...
+
+private binder = new Binder(this, EmployeeModel);
+
+render() {
+  const { model } = this.binder;
+
+  return html`
+    <vaadin-text-field label="Username"
+      ...=${field(model.username)}></vaadin-text-field>
+    <vaadin-text-field label="Title"
+      ...=${field(model.title)}></vaadin-text-field>
+    <vaadin-email-field label="Email"
+      ...=${field(model.email)}></vaadin-email-field>
+  `;
+}
+----
+
+image:images/bean-validation-errors.gif[The validation errors]
+
+=== Marking Fields as Required
+
+Certain constraints mark fields as required, i.e. add a visual "required" indicator.
+The standard constraints that make fields required are `@NotNull`, `@NotEmpty` and `@Size` (if `min()` value is greater than 0).
+
+For custom constraints this is controlled by the `impliesRequired` property in the TypeScript `validator` object, as shown in the <<tutorial-binder-validation#single-field-ts-validators>> section below.
```

Comment:
The section below confusingly describes “custom constraints” but has little to do with TypeScript validator objects. Should probably be using a different term, or describe how to define a Java constraint that marks a field as required.